src/librustc_mir/interpret/eval_context.rs

```diff
@@ -212,6 +212,48 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> LayoutOf for InterpCx<'mir, 'tcx, M> {
     }
 }
 
+/// Test if it is valid for a MIR assignment to assign `src`-typed place to `dest`-typed value.
+/// This test should be symmetric, as it is primarily about layout compatibility.
+pub(super) fn mir_assign_valid_types<'tcx>(src: Ty<'tcx>, dest: Ty<'tcx>) -> bool {
+    src == dest
+        || match (&src.kind, &dest.kind) {
+            (ty::Ref(_, src_pointee, _), ty::Ref(_, dest_pointee, _)) => {
+                // After optimizations, there can be assignments that change reference mutability.
+                // This does not affect reference layout, so that is fine.
+                src_pointee == dest_pointee
+            }
+            (ty::FnPtr(_), ty::FnPtr(_)) => {
```

Comment:
I think we could erase regions and try comparing the result again. Although I do wonder about what @eddyb said, so we may end up in the `false` arm if the assignment is between two instances of `struct Foo<T>(T);` where `T` is used for two function pointers that only differ between their HKL.

Reply:
You can't trivially erase lifetimes bound by *higher-ranked types* (I don't think HKL means anything, there is no `'Foo<'a, 'b>`).

I should come up with an example showing lifetime depending on the shape of bound lifetimes.